File: domain/source.ts
/**
 * domain/source.ts
 *
 * Source is the pure domain model for an originating channel from which
 * Evidence is collected. A Source carries an independent reliability
 * assessment (a ConfidenceScore) that propagates into the confidence of
 * everything collected from it, per the Confidence Everywhere principle.
 *
 * Mirrors contracts/domain/source.schema.json field-for-field.
 *
 * Domain Purity: standard runtime and sibling domain modules only.
 */

import {
  Status,
  freezeTags,
  newId,
  now,
  type Id,
  type Metadata,
  type Provenance,
  type Timestamp,
  EMPTY_METADATA,
} from "./common.js";
import { ClassificationLevel } from "./classification_level.js";
import { ConfidenceScore } from "./confidence_score.js";

export enum SourceType {
  PublicRegistry = "public_registry",
  SearchEngine = "search_engine",
  SocialPlatform = "social_platform",
  LeakedDataset = "leaked_dataset",
  SensorFeed = "sensor_feed",
  HumanReport = "human_report",
  InternalSystem = "internal_system",
  Other = "other",
}

export interface SourceInit {
  readonly sourceType: SourceType;
  readonly name: string;
  readonly uri?: string;
  readonly reliability?: ConfidenceScore;
  readonly status?: Status;
  readonly classification?: ClassificationLevel;
  readonly provenance: Provenance;
  readonly tags?: ReadonlyArray<string>;
  readonly metadata?: Metadata;
}

/** Source is an immutable record of a channel Evidence can be collected from. */
export class Source {
  readonly id: Id;
  readonly version: string;
  readonly createdAt: Timestamp;
  readonly updatedAt: Timestamp;
  readonly status: Status;
  readonly classification: ClassificationLevel;
  readonly sourceType: SourceType;
  readonly name: string;
  readonly uri?: string;
  readonly reliability?: ConfidenceScore;
  readonly provenance: Provenance;
  readonly tags: ReadonlyArray<string>;
  readonly metadata: Metadata;

  private constructor(id: Id, version: string, createdAt: Timestamp, updatedAt: Timestamp, init: SourceInit) {
    if (init.name.trim().length === 0) {
      throw new Error("Source.name must not be empty.");
    }
    this.id = id;
    this.version = version;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.status = init.status ?? Status.Active;
    this.classification = init.classification ?? ClassificationLevel.Public;
    this.sourceType = init.sourceType;
    this.name = init.name;
    this.uri = init.uri;
    this.reliability = init.reliability;
    this.provenance = init.provenance;
    this.tags = freezeTags(init.tags);
    this.metadata = init.metadata ?? EMPTY_METADATA;
    Object.freeze(this);
  }

  static create(init: SourceInit): Source {
    const timestamp = now();
    return new Source(newId(), "1.0.0", timestamp, timestamp, init);
  }

  /** True if this source has no independent reliability assessment yet. */
  isUnassessed(): boolean {
    return this.reliability === undefined;
  }
}

File: domain/source.py
"""
domain/source.py

Source domain model - origin of collected evidence.
Mirrors contracts/domain/source.schema.json field-for-field.
"""
from dataclasses import dataclass, field
from enum import Enum

from .classification_level import ClassificationLevel, ClassificationLevelValue
from .confidence_score import ConfidenceScore, ConfidenceLevel, ConfidenceMethods
from .common import (



    EMPTY_METADATA,
    Metadata,
    Provenance,
    Reference,
    Status,
    new_id,
    utc_now,
)


class SourceType(str, Enum):
    """Types of sources for collected evidence."""
    WEB = "web"
    API = "api"
    DATABASE = "database"
    DOCUMENT = "document"
    USER_INPUT = "user_input"
    AUTOMATED = "automated"
    FEED = "feed"
    OTHER = "other"


class SourceReliability(str, Enum):
    """Reliability rating for a source."""
    UNKNOWN = "unknown"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    VERY_HIGH = "very_high"


@dataclass(frozen=True)
class Source:
    """
    Source is the origin of collected evidence.
    
    Every Evidence item traces back to a Source, which carries
    provenance and reliability metadata.
    """
    id: str
    version: str
    created_at: str
    updated_at: str
    source_type: SourceType
    name: str
    url: str | None
    status: Status
    classification: ClassificationLevel
    reliability: SourceReliability
    collected_at: str
    collected_by: str
    contact_info: str | None
    provenance: Provenance
    tags: frozenset = field(default_factory=frozenset)
    metadata: Metadata = field(default_factory=lambda: dict(EMPTY_METADATA))

    @classmethod
    def create(
        cls,
        source_type: SourceType,
        name: str,
        collected_by: str,
        provenance: Provenance,
        collected_at: str | None = None,
        url: str | None = None,
        status: Status = Status.ACTIVE,
        classification: ClassificationLevelValue = ClassificationLevelValue.INTERNAL,
        reliability: SourceReliability = SourceReliability.UNKNOWN,
        contact_info: str | None = None,
        tags: list[str] | None = None,
        metadata: Metadata | None = None,
    ) -> "Source":
        """Create a new Source."""
        timestamp = utc_now()
        return cls(
            id=new_id(),
            version="1.0.0",
            created_at=timestamp,
            updated_at=timestamp,
            source_type=source_type,
            name=name,
            url=url,
            status=status,
            classification=ClassificationLevel(classification),
            reliability=reliability,
            collected_at=collected_at or timestamp,
            collected_by=collected_by,
            contact_info=contact_info,
            provenance=provenance,
            tags=frozenset(tags) if tags else frozenset(),
            metadata=dict(metadata) if metadata else dict(EMPTY_METADATA),
        )


__all__ = [
    "Source",
    "SourceReliability",
    "SourceType",
]
